Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,135 @@
+  | otherwise        = findSolution xs matrix result
+  where
+    result = matrix `mult` vector
+
+-- This routine takes a polynomial p, a matrix A and a vector w and
+-- returns p(A)w. It assumes the first coefficient of p is non zero,
+-- in particular that it is non empty. This makes the implementation
+-- easier as there is no need to write the identity matrix.
+evaluate :: SBMatrix -> SBVector -> [Bit] -> SBVector
+evaluate matrix w = foldr (\coeff acc -> (matrix `mult` acc) <> (if unBit coeff then w else mempty)) mempty
+
+findCandidatePoly :: SBMatrix -> SBVector -> SBVector -> F2Poly
+findCandidatePoly matrix z x = berlekampMassey dim errorPoly randomSequence
+  where
+    randomSequence = generateData matrix z x
+    errorPoly = fromInteger (1 `shiftL` (2 * dim)) :: F2Poly
+    dim = size matrix
+
+berlekampMassey :: Int -> F2Poly -> F2Poly -> F2Poly
+berlekampMassey dim = go 1 0
+  where
+    -- Is there a better way to implement recursion in this situation?
+    go :: F2Poly -> F2Poly -> F2Poly ->F2Poly -> F2Poly
+    go oneBefore twoBefore a b
+      | U.length (unF2Poly b) <= dim = oneBefore
+      -- Updated value is given by @twoBefore - oneBefore * q@
+      | otherwise                    = go (twoBefore - oneBefore * q) oneBefore b r
+        where
+          (q, r) = quotRem a b
+
+-- The input is a matrix B and a random vector z
+generateData :: SBMatrix -> SBVector -> SBVector -> F2Poly
+generateData matrix z x = toF2Poly $ U.fromList $ reverse $ map (x `dot`) matrixPowers
+  where
+    matrixPowers = L.take (2 * size matrix) $ L.iterate (matrix `mult`) z
+    -- Make sure x is not empty.
+    -- Check matrix indices
+    -- trace ("Size of x: " ++ (show ((I.foldr (\entry acc -> acc + (if entry == mempty then 0 else 1)) 0 ix) :: Int))) $
+
+-- Infinite lists of random vectors
+getRandomVectors :: [Int] -> Double -> StdGen -> [SBVector]
+getRandomVectors rows sparsity gen = go randomEntries
+  where
+    randomEntries = zip (cycle rows) (randomRs (0, 1) gen)
+    go :: [(Int, Double)] -> [SBVector]
+    go list = newVector : go backOfList
+      where
+        newVector = SBVector (S.fromList listOfEntries)
+        listOfEntries = fmap fst $ filter (\(_, rDouble) -> rDouble < sparsity) frontOfList
+        (frontOfList, backOfList) = L.splitAt (length rows) list
+
+-- Informal way of testing large matrices
+-- Input number of columns of matrix and sparsity coefficient
+testLinearSolver :: Int -> Double -> Bool
+testLinearSolver dim sparsity = mat `mult` sol == mempty
+  where
+    sol = linearSolve mat
+    mat = SBMatrix (V.fromList listOfColumns)
+    -- -2 is arbitrary. It means that the number of rows is at most one less than
+    -- the number of columns
+    listOfColumns = L.take dim $ getRandomVectors [0..(dim - 2)] sparsity $ mkStdGen $ fromIntegral $ unsafePerformIO getMonotonicTimeNSec
+
+
+-- 1) Investigate rare cases where code does not terminates.
+-- 2) Implement dense matrices rather than sparse.
```

Comment:
It could be interesting to use dense matrices to speed up multiplication, but I don't quite understand how to fit it into memory. In our toy example we have 5621 equations over 1489 primes, which is 8М bits and 1M bytes. This is easily doable. But AFAIU it is readily expected to have 1000 times more equations, isn't it? What number of equations is quoted in references?

Reply:
I am quoting from the slides.
Researchers broke a 768 bits RSA key using NFS with Wiedemann's algorithm. The input binary matrix had 200 millions of rows and 150 non zeros elements by rows. It took 98 days of computations on a cluster of 576 cores.

Reply:
In this case sparse representation would take 200*10^6 * 150 * 8 = 240 Gb, which is feasible for a cluster of their size. But dense representation would take something like 1250000 Gb, which is far beyond current capabilities.